packages/wallet/src/signer.ts

```diff
@@ -1,27 +1,114 @@
-import { AminoSignResponse, OfflineAminoSigner, Secp256k1Wallet, StdSignDoc } from "@cosmjs/amino";
+import { AminoSignResponse, makeSignDoc as makeSignDocAmino, OfflineAminoSigner, Secp256k1Wallet, StdSignDoc } from "@cosmjs/amino";
 import { LedgerSigner } from "@cosmjs/ledger-amino";
-import { AccountData, DirectSecp256k1Wallet, DirectSignResponse, OfflineDirectSigner } from "@cosmjs/proto-signing";
-import { constructAdr36SignDoc } from "@demex-sdk/core";
+import { AccountData, DirectSecp256k1Wallet, DirectSignResponse, makeSignDoc as makeSignDocProto, OfflineDirectSigner, TxBodyEncodeObject } from "@cosmjs/proto-signing";
+import { StdFee } from "@cosmjs/stargate";
+import { AminoTypesMap } from "@demex-sdk/amino-types";
+import { registry } from "@demex-sdk/codecs";
+import { AuthInfo, TxBody } from "@demex-sdk/codecs/data/cosmos/tx/v1beta1/tx";
+import { constructAdr36SignDoc, evmChainIds } from "@demex-sdk/core";
 import { SignDoc } from "cosmjs-types/cosmos/tx/v1beta1/tx";
-import { EIP712Tx } from "./eip712";
 import { WalletError } from "./constant";
-
-export enum DemexSignerTypes {
-  Ledger,
-  PrivateKey,
-  PublicKey,
-}
+import { constructEIP712Tx, EIP712Tx, parseChainId } from "./eip712";
+import { DemexSignerTypes, getEvmHexAddress } from "./utils";
 export interface EIP712Signer {
-  getEvmAddress: () => Promise<string>;
   getEvmChainId: () => Promise<string>;
   signEIP712: (hexAddress: string, doc: EIP712Tx) => Promise<string>;
 }
 
-export type DemexEIP712Signer = (DemexDirectSigner | DemexAminoSigner) & EIP712Signer
 export type DemexSigner = DemexDirectSigner | DemexAminoSigner | DemexEIP712Signer;
 export type DemexDirectSigner = OfflineDirectSigner & { type: DemexSignerTypes };
 export type DemexAminoSigner = OfflineAminoSigner & { type: DemexSignerTypes }
 
+export abstract class DemexEIP712Signer implements DemexDirectSigner, DemexAminoSigner, EIP712Signer {
+
+  type = DemexSignerTypes.EIP712;
+
+  abstract getAccounts(): Promise<readonly AccountData[]>;
+  abstract signEIP712(hexAddress: string, doc: EIP712Tx): Promise<string>;
+  abstract getEvmChainId(): Promise<string>;
+
+  async signAmino(address: string, signDoc: StdSignDoc): Promise<AminoSignResponse> {
+    const walletChainId = await this.getEvmChainId()
+    const pubkey = await this.getPublicKey(address)
+    const hexAddress = getEvmHexAddress(pubkey)
+    const evmChainId = evmChainIds[signDoc.chain_id]!
```

Comment:
why are we sure `evmChainIds` contains the `signDoc.chain_id`? maybe a check and throw is more appropriate.